src/main/java/net/jonathangiles/tools/teenyhttpd/implementation/TypedResponseImpl.java

```diff
@@ -0,0 +1,91 @@
+package net.jonathangiles.tools.teenyhttpd.implementation;
+
+import net.jonathangiles.tools.teenyhttpd.model.Header;
+import net.jonathangiles.tools.teenyhttpd.model.StatusCode;
+import net.jonathangiles.tools.teenyhttpd.model.TypedResponse;
+
+import java.io.BufferedOutputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public final class TypedResponseImpl<T> implements TypedResponse<T> {
+
+    private final StatusCode statusCode;
+    private T body;
+    private final Map<String, List<String>> headers;
```

Comment:
I wonder if we can internally use the same code for headers as in Response, where it is a Map<String, Header>?

Reply:
I would like to replace the current Header class with:

```java

import java.util.*;

public class Headers extends HashMap<String, List<String>> {
    
    public Headers add(String key, String value) {
        putIfAbsent(key, new ArrayList<>());
        get(key).add(value);
        return this;
    }

    public Headers addAll(String key, String... values) {
        putIfAbsent(key, new ArrayList<>());
        get(key).addAll(Arrays.asList(values));
        return this;
    }

    public Headers addAll(String key, List<String> values) {
        putIfAbsent(key, new ArrayList<>());
        get(key).addAll(values);
        return this;
    }

    public Headers set(String key, String value) {
        put(key, new ArrayList<>(Collections.singleton(value)));
        return this;
    }

    public Headers setAll(String key, String... values) {
        put(key, new ArrayList<>(Arrays.asList(values)));
        return this;
    }

    public Headers setAll(String key, List<String> values) {
        put(key, new ArrayList<>(values));
        return this;
    }

    public Headers remove(String key) {
        super.remove(key);
        return this;
    }

}

```

The current  Map<String, Header> is prune to errors since this structure allows the next case:

```
Map<String, Header> headers = new HashMap<>();
headers.put("Authorization", new Header("Content-Type", List.of("Bearer 1234"));
```

Reply:
The original intent of the Header class was to defer the parsing of incoming request headers until later. That is why it takes a `String line` constructor as one variant. I don't think there is any code in the existing API that can lead to that error scenario (from a users perspective), but correct me if I'm wrong.

Reply:
I believe we must parse the headers eagerly in order to sanitize them and made validations before hitting users code, for example, validate the `Content-Type` header and reject the request if no proper `MesssageConverter` are available, validate ETags, or simply validate the header `Host` don't appear twice.


Following headers aren't validated until you need it.

```
Host: google.com,facebook.com
Authorization:
```

but for now I'll use the current Header class

Reply:
Response doesn't have a Map<String, Header>, but Request does :S